lib/rack/request.rb

```diff
@@ -18,7 +18,15 @@ class << self
     end
 
     self.ip_filter = lambda { |ip| ip =~ /\A127\.0\.0\.1\Z|\A(10|172\.(1[6-9]|2[0-9]|30|31)|192\.168)\.|\A::1\Z|\Afd[0-9a-f]{2}:.+|\Alocalhost\Z|\Aunix\Z|\Aunix:/i }
-    SCHEME_WHITELIST = %w(https http).freeze
+    ALLOWED_SCHEMES = %w(https http).freeze
+    SCHEME_WHITELIST = ALLOWED_SCHEMES
+    if $VERBOSE
+      if Object.respond_to?(:deprecate_constant)
+        deprecate_constant :SCHEME_WHITELIST
+      else
+        warn("SCHEME_WHITELIST is deprecated and will be removed in a future version of Rack. Please use ALLOWED_SCHEMES instead.")
```

Comment:
I just realized that this will always warn even if `SCHEME_WHITELIST` is not used by the app. Is that desired?

Reply:
I did some experiments:

```ruby
class Namespace
  ALLOWED_SCHEMES = %w(http https)

  SCHEME_WHITELIST = ALLOWED_SCHEMES
  if $VERBOSE
    if Object.respond_to?(:deprecate_constant)
      deprecate_constant :SCHEME_WHITELIST
    else
      warn("SCHEME_WHITELIST is deprecated. Please use ALLOWED_SCHEMES instead.")
    end
  end
end

puts "Reference ALLOWED_SCHEMES"
Namespace::ALLOWED_SCHEMES

puts "Reference SCHEME_WHITELIST"
Namespace::SCHEME_WHITELIST
```

=>


```
$ ruby -w namespace.rb
Reference ALLOWED_SCHEMES
Reference SCHEME_WHITELIST
1.rb:19: warning: constant Namespace::SCHEME_WHITELIST is deprecated
```

Only when we actually reference `SCHEME_WHITELIST` and `$VERBOSE` is on will emit warnings, loading the file will not emit warnings.

Reply:
But is not that because `deprecate_constant` is defined? If you don't have it it happens as soon that like 27 is executed

Reply:
Ah I understand now. Yes, for Ruby < 2.3.0, if they turn on warning (`-w`, `--verbose`, `RUBYOPT=-w`), then a warning will show as soon as L27 executed. What do you suggest?

Reply:
I'd say, let's remove the else branch for now. There is a chance that rack master will not support Ruby 2.2